src/sqlfluff/core/linter.py

```diff
@@ -105,275 +101,136 @@ def is_clean(self):
+        untouchable_slices = self.templated_file.untouchable_slices()
+
+        linter_logger.debug("Untouchables: %s", untouchable_slices)
+
+        # Patches, sorted by start
+        template_space_patches: List[FixPatch] = sorted(
+            list(
+                self.tree.iter_patches(templated_str=self.templated_file.templated_str)
+            ),
+            key=lambda x: x[0].start,
         )
-        bencher("fix_string: Matched 1&2")
-        # diff_fix = SequenceMatcher(autojunk=None, a=self.file_mask[1][0], b=self.file_mask[2][0])
-        diff_fix_codes = diff_fix.get_opcodes()
-        linter_logger.debug("Fixing diff codes: %s", diff_fix_codes)
-        bencher("fix_string: Got Opcodes 1&2")
-
-        # If diff_templ isn't the same then we should just keep the template. If there *was*
-        # a fix in that space, then we should raise an issue
-        # If it is the same, then we can apply fixes as expected.
-        write_buff = ""
-        fixed_block = None
-        templ_block = None
-        # index in raw, templ and fix
-        idx = (0, 0, 0)
-        loop_idx = 0
-        bencher("fix_string: Loop Setup")
-        while True:
-            loop_idx += 1
-            linter_logger.debug(
-                "%04d: Write Loop: idx:%s, buff:%r", loop_idx, idx, write_buff
-            )
-            if templ_block is None:
-                if diff_templ_codes:
-                    templ_block = diff_templ_codes.pop(0)
-                # We've exhausted the template. Have we exhausted the fixes?
-                elif fixed_block is None and not diff_fix_codes:
-                    # Yes - excellent. DONE
-                    break
-                # Deal with the case that we only have inserts left.
-                elif all(elem[0] == "insert" for elem in diff_fix_codes):
-                    for fixed_block in diff_fix_codes:
-                        write_buff += self.file_mask[2][fixed_block[3] : fixed_block[4]]
-                    break
-                else:
-                    raise NotImplementedError(
-                        "Fix Block(s) left over! Don't know how to handle this! aeflf8wh"
-                    )
-            if fixed_block is None:
-                if diff_fix_codes:
-                    fixed_block = diff_fix_codes.pop(0)
-                elif templ_block[0] != "delete":
-                    # We need another fixed_block for the cases where templ_block[0] is not 'delete'
-                    raise NotImplementedError(
-                        "A {} template block remains with no more diff_fix_codes left".format(
-                            templ_block[0]
-                        )
-                    )
+        linter_logger.debug("Templated-space patches: %s", template_space_patches)
 
-            linter_logger.debug(
-                "%04d: Blocks: template:%s, fix:%s", loop_idx, templ_block, fixed_block
+        # We now convert enrich the patches into source space
```

Comment:
Can you clarify that comment? Does that mean we're applying the patches?

Reply:
Good feedback. I've edited this comment to `We now convert templated-space patches into source-space patches.`.

What we're doing is taking a load of patches which are referenced using positions in the templated file to ones which are referred to in the source file, which means we can dedupe, and also apply them around templated sections.